lgc/builder/BuilderImpl.cpp

```diff
@@ -74,6 +74,161 @@ Value *BuilderImplBase::CreateDotProduct(Value *const vector1, Value *const vect
   return scalar;
 }
 
+// =====================================================================================================================
+// Create code to calculate the dot product of two integer vectors, with optional accumulator, using hardware support
+// where available. The factor inputs are always <N x iM> of the same type, N can be arbitrary and M must be 4, 8, 16,
+// 32, or 64 Use a value of 0 for no accumulation and the value type is consistent with the result type. The result is
+// saturated if there is an accumulator.
```

Comment:
Only the final addition to the accumulator needs to be saturated. Intermediate overflows of the dot product can lead to an undefined result.